src/D2L.CodeStyle.Analyzers/Immutability/MutabilityInspectionResult.cs

```diff
@@ -23,14 +23,14 @@ public enum MutabilityCause {
 
 	public sealed class MutabilityInspectionResult {
 
-		private readonly static MutabilityInspectionResult s_notMutableResult = new MutabilityInspectionResult( false, null, null, null, null, ImmutableHashSet<string>.Empty );
-
 		public bool IsMutable { get; }
 
 		public string MemberPath { get; }
 
 		public string TypeName { get; }
 
+		public ISymbol Symbol { get; }

```

Comment:
What is `Symbol`?

`MutabilityInspectionResult` is hellla confusing because its two cases merged into one (mutable vs. not mutable) but things like `TypeName` are self-explanitory if you notice its only used for `Mutable` (in which case its storing data for a particular reason the class is mutable (sadly we only return an arbitrary reason, not all the reasons.))

`Symbol` also seems to be being used for the "not mutable" case which is surprising - there are a lot of symbols that occur in the process of judging something as not mutable so why do we care about a particular one?

I haven't read the rest of the diff yet but this seems fishy. If it makes sense maybe we need to document it in a comment.

Reply:
> What is Symbol?

I don't have the clearest understanding, I was actually going to ask you. My understanding is that is how the compiler represents things like `namespace, class, method, parametenamespace, class, method, parameter` (got that from the documented in the ISymbol interface). I believe it is anything we are interested in analyzing.

> Symbol also seems to be being used for the "not mutable" case which is surprising

I interpret it as after inspecting symbol `s` we found it to be immutable. I think of an inspector walking down some path and taking notes on every symbol it analyzes, it then marks each symbol as either mutable or immutable.  We want to have the symbol for the "not mutable" case because what if we happen to stumble on a `Mutability.Unaudited` and find out it is Immutable. We need to know that specific symbol was immutable, and there might be more than one, also mixed in with real violations.


> there are a lot of symbols that occur in the process of judging something as not mutable so why do we care about a particular one?

Perhaps there are some optimizations to be made once we start scanning past `Unaudited` and `Audited`, where we will only keep the mutability violators, and annotated violators and throw out the ones that are fine. But we need the care of a particular one because unlike before where we returned what could be thought of as one inspection result, soon we will be able to return an aggregate of inspection results.


We can talk tomorrow if I'm off or there is a better way.

Reply:
Yes my original description is wrong `TypeName` must stay but `MemberPath` can be derived from the ISymbol

Reply:
> >   What is Symbol?
>
> I don't have the clearest understanding, I was actually going to ask you. My understanding is that is how the compiler represents things like namespace, class, method, parametenamespace, class, method, parameter (got that from the documented in the ISymbol interface). I believe it is anything we are interested in analyzing.

So `ISymbol` (not "Symbol") identifies a semantic object, that I'm cool with. What I was asking though is what is `MutabilityInspectorResult.Symbol` - how would you describe it? But that was addressed next:

> I interpret it as after inspecting symbol s we found it to be immutable. I think of an inspector walking down some path and taking notes on every symbol it analyzes, it then marks each symbol as either mutable or immutable. We want to have the symbol for the "not mutable" case because what if we happen to stumble on a Mutability.Unaudited and find out it is Immutable. We need to know that specific symbol was immutable, and there might be more than one, also mixed in with real violations.

Ok - so this is were I'm not following. There would be multiple symbols in that case, e.g.

```cs
[Mutability.Unaudited( ... )]
private readonly SafeThing m_safe;

// ...

public sealed class SafeThing {
    public readonly int m_immutableThing1;
    public readonly int m_immutableThing2;
}
```

`MutabilityInspectionResult` can only hold one symbol. Is your goal for it to hold the symbol for `SafeThing`, not `m_immutableThing1` or `m_immutableThing2` when we return from inspecting `SafeThing`? What about when we return from inspecting the `m_safe` field - should it be `m_safe` or `SafeThing`?

Reply:
My intention is when inspecting `SafeThing` you get symbol for `SafeThing`, when inspecting `m_safe` you get symbol for `m_safe`. Internally it is a little more interesting than that because our current implementation for say `InspectField`, doesn't only verify if the field is mutable, it will also verify if its type is mutable. This means if you inspect `m_safe` you will get a symbol for `m_safe` if it is not readonly. But if when inspect `m_safe` it is `readonly` then you will get a symbol for `SafeThing` because that is the final thing you inspect.  That is how it would work now.

In the future I want to surface all results from the inspection. You could think of the current `MutabilityInspectionResult` to hold only one result, where I'd like to hold multiple results. You can think of it being restructured to as roughly follows (some fields omitted).

After: 
```
{  
  MutabilityInspectionResult:{  
    IsMutable: false,
    UserMarkedIgnored: true,
    Symbol: m_safe,
    Children:[  
      {  
        MutabilityInspectionResult:{  
          IsMutable:false,
          UserMarkedIgnored: false,
          Symbol: SafeThing,
          Children:[  
              {  
                MutabilityInspectionResult:{  
                  IsMutable:false,
                  UserMarkedIgnored: false,
                  Symbol:m_immutableThing1,
                  Children:[  

                  ]
                }
              },
              {  
                MutabilityInspectionResult:{  
                  IsMutable:false,
                  UserMarkedIgnored:false,
                  Symbol:m_immutableThing2,
                  Children:[  

                  ]
                }
              }
          ]
        }
      }
    ]
  }
}
```

Before:
```
{
  MutabilityInspectionResult: {
    IsMutable: false
  }
}
```